### ISSUE TITLE
Null objects in returned list when all properties are null

### ISSUE THREAD

**Issue by miqm** [dotnet/efcore#21308]
I'm using DBModelFirst approach.

If we have a table or a view, that has no key and all properties can be null and if there is a row that all properties are null, EF core does not return an object with all properties being nulls but a null object itself.

This can be very misleading, especially when used with nullable references.

Although Visual studio hints on the foreach shows that "item" is "TestItem?", when used without ifs shows that "item is not null here".

![image](https://user-images.githubusercontent.com/7921224/84886906-558ad900-b095-11ea-8f0b-e01605b8db62.png)
![image](https://user-images.githubusercontent.com/7921224/84886926-5d4a7d80-b095-11ea-8b17-c87ec579836b.png)
![image](https://user-images.githubusercontent.com/7921224/84887112-971b8400-b095-11ea-91c1-8aeaf3fec38b.png)

However if we'd expect items in the projected DbSet to be nulls, we should have `List<TestTableItem?>`.

### Steps to reproduce

``` SQL
CREATE TABLE [dbo].[TestTable](
	[Prop1] [nchar](10) NULL,
	[Prop2] [nchar](10) NULL,
	[Prop3] [nchar](10) NULL,
	[Prop4] [nchar](10) NULL
)
GO

CREATE VIEW [dbo].[TestView]
AS
SELECT dbo.TestTable.* FROM  dbo.TestTable
GO

INSERT INTO [dbo].[TestTable]
           ([Prop1]
           ,[Prop2]
           ,[Prop3]
           ,[Prop4])
     VALUES
	   ('11' ,NULL ,NULL ,NULL)
          ,(NULL ,'22' ,NULL ,NULL)
	  ,(NULL ,NULL ,'33' ,NULL)
	  ,(NULL ,NULL ,NULL ,'44')
	  ,(NULL ,NULL ,NULL ,NULL)
GO
```
``` C#
class Program
    {

        static async Task Main(string[] args)
        {
            var optB = new DbContextOptionsBuilder<Context>().UseSqlServer(Opts.ConnectionString);
            var ctx = new Context(optB.Options);
            {
                Console.WriteLine($"TestTable:");
                var items = await ctx.TestTable.ToListAsync();
                foreach (var item in items)
                {
                    if (item is null)
                        Console.WriteLine($"null");
                    else
                        Console.WriteLine($"{item.Prop1} | {item.Prop2} | {item.Prop3} | {item.Prop4}");
                }
            }
            Console.WriteLine($"--------");
            {
                Console.WriteLine($"TestView:");
                var items = await ctx.TestView.ToListAsync();
                foreach (var item in items)
                {
                    if (item is null)
                        Console.WriteLine($"null");
                    else
                        Console.WriteLine($"{item.Prop2} | {item.Prop3}");
                }
            }
        }
    }

    class Context : DbContext
    {
        public Context(DbContextOptions<Context> options) : base(options)
        {

        }
        public virtual DbSet<TestTableItem> TestTable { get; set; } = null!;
        public virtual DbSet<TestViewItem> TestView { get; set; } = null!;

        protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            base.OnModelCreating(modelBuilder);
            modelBuilder.Entity<TestTableItem>().HasNoKey();
            modelBuilder.Entity<TestViewItem>().HasNoKey();
        }
    }

    [Table("TestView", Schema = "dbo")]
    public class TestViewItem
    {
        public string? Prop2 { get; set; }
        public string? Prop3 { get; set; }
    }

    [Table("TestTable", Schema = "dbo")]
    public class TestTableItem
    {
        public string? Prop1 { get; set; }
        public string? Prop2 { get; set; }
        public string? Prop3 { get; set; }
        public string? Prop4 { get; set; }
    }
```

### Further technical details

EF Core version: 3.1.5
Database provider: Microsoft.EntityFrameworkCore.SqlServer
Target framework: .NET Core 3.1
Operating system: Windows 10
IDE: Visual Studio 2019 16.6.2 & Visual Studio 2019 16.7.0-preview2

**Comment by smitpatel** [comment]
Duplicate of #19537

**Comment by miqm** [comment]
Damn, I swear I was looking through the issues for this one :)

Confirmed, that on 5.0-preview5 works properly.
Will this be fixed also in 3.1.x? Seems like this is a quite big bug that should be fixed in LTS channel.

**Comment by ajcvickers** [comment]
@miqm We discussed patching (see the [planning process](https://docs.microsoft.com/en-us/ef/core/what-is-new/release-planning)) but since this is a significant change in behavior we're not going to do this in a patch. Indeed, I added `breaking change` to the duplicate issue so that we document this change for 5.0.
